### DISCUSSION TITLE
How to export the projection

### DISCUSSION THREAD

**Discussion by ittayd** [khalim19/gimp-plugin-export-layers#112] [Q&A] [answered]
I'm so new to Gimp that I'm not even sure I'm using the right terms. I have a bunch of layers (photos) in a group and the group has a circle mask. I want to export the layers with this mask, so the result is a square image with only a circle of content. I think this is called 'projection'. 
Alternatively, I want to apply a mask before the export (that is, I don't care if the mask is applied in Gimp, I need it just for the result output)

**Comment by ittayd** [comment]
I tried to apply a procedure and got an error:
`Failed to apply procedure "gimp-layer-add-mask" on "2023_09_07_181805.jpg" because:`

`Procedure 'gimp-layer-add-mask' has been called with an invalid ID for argument 'mask'. Most likely a plug-in is trying to work on a layer that doesn't exist any longer.`

I tried to use a layer outside of the group, same results.

**Comment by kamilburda** [comment]
I found a solution, although it might be somewhat inconvenient as it involves multiple steps.

I created a Python script that applies a layer mask from a layer specified by its name to the current layer.

1. Install the script (link below) the same way as Export Layers by copying it to the `plug-ins` directory.
3. Switch to batch editing (Settings -> Batch Editing)
4. Add the `python-fu-apply-mask-from-layer` procedure and specify the name of the layer with the mask
5. Click Run
6. Switch back to the export mode by unchecking Settings -> Batch Editing
7. Uncheck `python-fu-apply-mask-from-layer`
8. Export

If you do not wish to keep the masks, you can simply undo the change.

The masks will not be created/overridden if a layer already has one.

[apply_mask_from_layer.zip](https://github.com/kamilburda/gimp-export-layers/files/14151987/apply_mask_from_layer.zip)

**Reply by ittayd** [reply]
Thank you. I copied the file to ~/.var/app/org.gimp.GIMP/config/GIMP/2.10/plug-ins/ (i have a flatpak installation) but after restarting gimp I can't find it in the custom procedures. Gimp is 2.10.36

**Comment by kamilburda** [comment]
If you use a Unix-based system, make sure the .py file is executable.

You should be able to see a new menu entry under Layer -> Apply mask from layer.

Also maybe check if the `plug-ins` folder is added in GIMP (Edit -> Preferences -> Folders -> Plug-ins).

**Reply by ittayd** [reply]
Yep, the executable bit was it. Thanks!

**Comment by ittayd** [comment]
I created a layer called Mask and added the mask to it (it's icon is a black rectangle with a white circle). Didn't apply the mask. The batch operation iterates over all layers and applies a mask, but apparently it's a 'do nothing' as it is all white (white rectangle in the icon).

**Reply by kamilburda** [reply]
The mask is not scaled and is pasted in the same offsets as the layers it is applied onto.

One thing I forgot to mention when batch-editing (before step 4) is to uncheck `Use layer size` as that will modify the image size every time another layer is being processed.

If that doesn't help, maybe you could send an example of layers and a mask if you're OK with it,

**Reply by ittayd** [reply]
All my layers are of the same size. 

Attached is an example, 
[example.zip](https://github.com/kamilburda/gimp-export-layers/files/14152578/example.zip)

**Reply by kamilburda** [reply, marked as answer]
When I cleared the selection (before calling the script), the mask was copied with the black background and the white circle (which I assume is the expected result). The script assumes that the selection is empty, which I did not anticipate. If you need to preserve the selection, you can create a new channel from selection (Select -> Save to Channel).

**Reply by ittayd** [reply]
Works like a charm! Thanks!
